factom_keys/id.py

```diff
@@ -0,0 +1,197 @@
+        ), "Only provide one of seed_bytes or key_string, not both"
+
+        if key_string:
+            if not IDPrivateKey.is_valid(key_string):
+                raise BadKeyStringError
+            decoded = base58.decode(key_string)
+            seed_bytes = decoded[PREFIX_LENGTH:BODY_LENGTH]
+
+        assert isinstance(seed_bytes, bytes)
+        assert len(seed_bytes) == 32
+        self.__signer = ed25519.SigningKey(seed_bytes)
+
+    @property
+    def key_bytes(self):
+        """
+        :return: the 32 byte raw private key
+        """
+        return self.__signer.to_seed()
+
+    def to_string(self, key_num: int):
+        """
+        :param key_num: the number of the ID key pair: x in (skx, idx) where 0 < x < 5
+        :return: the ID private key as a human-readable string in skx format
+        """
+        # Key number must be 1-4 for sk1-sk4
+        assert 0 < key_num < 5, "Key number must be 1-4 for keys sk1-sk4."
+
+        secret_body = IDPrivateKey.PREFIXES[key_num - 1] + self.key_bytes
+        return _to_base58_string(secret_body)
+
+    def get_public_key(self):
+        """
+        Derive and return the corresponding IDPublicKey
+
+        :return: the IDPublicKey corresponding to this IDPrivateKey
+        """
+        public_bytes = self.__signer.get_verifying_key().to_bytes()
+        return IDPublicKey(public_bytes)
+
+    def sign(self, message):
+        """
+        :param message: the bytes to sign
+        :return: the 64 byte signature of the message
+        """
+        return self.__signer.sign(message)
+
+    @classmethod
+    def is_valid(cls, key_string: str):
+        """
+        :param key_string: the ID private key string to be checked
+        :return: `True` if `key_string` is a valid ID private key string; `False` otherwise
+        """
+        if not isinstance(key_string, str):
+            return False
+
+        try:
+            decoded = base58.decode(key_string)
+        except base58.InvalidBase58Error:
+            return False
+
+        if len(decoded) != TOTAL_LENGTH or decoded[:PREFIX_LENGTH] not in IDPrivateKey.PREFIXES:
+            return False
+
+        checksum_claimed = decoded[BODY_LENGTH:]
+        temp_hash = sha256(decoded[:BODY_LENGTH]).digest()
+        checksum_actual = sha256(temp_hash).digest()[:CHECKSUM_LENGTH]
+
+        return checksum_actual == checksum_claimed
+
+
+class IDPublicKey:
+
+    # Prefixes for id1 - id4
+    PREFIXES = [b"\x3f\xbe\xba", b"\x3f\xbe\xd8", b"\x3f\xbe\xf6", b"\x3f\xbf\x14"]
+
+    def __init__(self, key_bytes=None, key_string=None):
+        assert (key_bytes and not key_string) or (
+            not key_bytes and key_string
+        ), "Only provide one of key_bytes or key_string, not both"
+
+        if key_string:
+            if not IDPublicKey.is_valid(key_string):
+                raise BadKeyStringError()
+            decoded = base58.decode(key_string)
+            key_bytes = decoded[PREFIX_LENGTH:BODY_LENGTH]
```

Comment:
When using a `key_string` rather than bytes, let's have this call the `set_key_level()` function i recommended, using the appropriate level corresponding to key string.

When using the `key_bytes`, we can have it default to level 4 (least powerful key), and they'll just have to call this function manually after creation if they want to use a different level.

Reply:
Makes sense. This made me realize I need to add some string or regex checking to the `is_valid` methods to ensure we get the prefixes we expect.